include/pybind11/pybind11.h

```diff
@@ -341,9 +340,18 @@ class cpp_function : public function {
                         std::string(pybind11::str(rec->scope.attr("__name__"))) + "." + std::string(rec->name) + signature
                     #endif
                 );
-            while (chain->next)
-                chain = chain->next;
-            chain->next = rec;
+
+            if (rec->is_prepended) {
+                // Beginning of chain
+                chain_start = rec;
+                rec->next = chain;
```

Comment:
I've pushed a commit to your branch that should get this working.  (Feel free to merge into your own commit). Description below.

The missing piece is that the current implementation is only half-updating the chain; `chain_start` is only used to generated the docstring, but the actual dispatch begins at the pointer stored inside the `rec_capsule` created by the code path when there is no existing chain:
initial definition a few lines above:
```C++
            capsule rec_capsule(rec, [](void *ptr) { // ...
            // ...

            m_ptr = PyCFunction_NewEx(rec->def, rec_capsule.ptr(), scope_module.ptr());
```
This `rec_capsule`, which points to the initial function, gets passed to the `dispatcher()` function as `self`; the dispatcher then starts the chain from that function record.

So in addition to setting `chain_start`, we're going to have to dig into that capsule and replace its pointer (but *without* destructing the capsule because destruction frees the chained function records).  This should do it:
```C++
            if (rec->is_prepended) {
                // Beginning of chain; we need to replace the capsule's current head-of-the-chain
                // pointer with this one, then make this one point to the previous head of the
                // chain.
                chain_start = rec;
                rec->next = chain;
                auto rec_capsule = reinterpret_borrow<capsule>(((PyCFunctionObject *) m_ptr)->m_self);
                rec_capsule.set_pointer(rec);
            } else {
```
which needs a new method `capsule::set_pointer` wrapping `PyCapsule_SetPointer`:

```C++
    // Replaces a capsule's pointer *without* calling the destructor on the existing one.
    void set_pointer(const void *value) {
        if (PyCapsule_SetPointer(m_ptr, const_cast<void *>(value)) != 0)
            pybind11_fail("Could not set capsule pointer");
    }
```